lib/streamlit/elements/arrow_table.py

```diff
@@ -0,0 +1,165 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018-2019 Streamlit Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pandas as pd
+import pyarrow as pa
+from streamlit import util
+from streamlit import type_util
+
+
+def marshall(proto, df, default_uuid):
+    if type_util.is_pandas_styler(df):
+        df = _marshall_styler(proto, df, default_uuid)
+
+    # Convert all dataframe values to strings as we
+    # only use them to show the result.
+    df = df.astype(str)
+
+    _marshall_headers(proto, df)
+    _marshall_data(proto, df)
+
+
+def _marshall_styler(proto, styler, default_uuid):
+    # NB: Order matters!
+    # UUID should be set before _compute is called.
+    _marshall_table_uuid(proto, styler, default_uuid)
+    _marshall_table_caption(proto, styler)
+
+    # NB: We're using protected members of Styler to get styles,
+    # which is non-ideal and could break if Styler's interface changes.
+    styler._compute()
+    pandas_styles = styler._translate()
+
+    _marshall_table_styles(proto, styler, pandas_styles)
+
+    old_df = styler.data
+    new_df = _use_display_values(old_df, pandas_styles)
+
+    return new_df
+
+
+def _marshall_table_uuid(proto, styler, default_uuid):
+    if styler.uuid is None:
+        styler.set_uuid(default_uuid)
+
+    proto.uuid = str(styler.uuid)
+
+
+def _marshall_table_caption(proto, styler):
+    proto.caption = styler.caption if styler.caption is not None else ""
+
+
+def _marshall_table_styles(proto, styler, styles):
+    css_rules = []
+
+    if "table_styles" in styles:
+        table_styles = styles["table_styles"]
+        table_styles = _trim_pandas_styles(table_styles)
+        for style in table_styles:
+            # NB: styles in "table_styles" have a space
+            # between the UUID and the selector.
+            rule = _pandas_style_to_css(style, styler.uuid, separator=" ")
+            css_rules.append(rule)
+
+    if "cellstyle" in styles:
+        cellstyle = styles["cellstyle"]
+        cellstyle = _trim_pandas_styles(cellstyle)
+        for style in cellstyle:
+            rule = _pandas_style_to_css(style, styler.uuid)
+            css_rules.append(rule)
+
+    proto.styles = "\n".join(css_rules)
+
+
+def _trim_pandas_styles(styles):
+    # Filter out empty styles, as every cell will have a class
+    # but the list of props may just be [['', '']].
+    return [x for x in styles if any(any(y) for y in x["props"])]
+
+
+def _pandas_style_to_css(style, uuid, separator=""):
```

Comment:
I wonder this could be done client side so that we don't have to pass a CSS string over proto. May not know enough here...

Reply:
What would you pass in the proto instead?

Reply:
The main idea is to construct the CSS from the "inner" structure of the dataframe. Theoretically, that can be done on the client side - but it will require passing the whole dict over the proto.